weaknesses/MASVS-CRYPTO/MASWE-0019.md

```diff
@@ -1,38 +1,51 @@
-    expert-designed APIs or certified modules exposes apps to vulnerabilities due
-    to potential implementation flaws and lack of rigorous security review.
-  topics:
-  - platform-provided cryptographic APIs (e.g. conscrypt/CryptoKit)
-  - custom-made cryptographic APIs (e.g. via xor, bit flipping, etc. or cryptographic
-    constants or values such as sbox, etc.)
-  - custom algorithms, primitives, protocols
-  - specify Cipher.getInstance provider (Android)
-  - Android Security Provider (Android)
-  - Jetpack Security Crypto Library (Android)
-  - BoucyCastle algorithms (Android)
-status: draft
-
----
-
+---
+title: Potentially Weak Cryptography Implementations
+alias: potentially-weak-crypto-impl
+platform: [android, ios]
+profiles: [L2]
+mappings:
+  masvs-v1: [MSTG-CRYPTO-2]
+  masvs-v2: [MASVS-CRYPTO-1, MASVS-CODE-3]
+  mastg-v1: [MASTG-TEST-0061, MASTG-TEST-0014]
+
+refs: 
+ - https://cwe.mitre.org/data/definitions/1240.html
+ - https://cwe.mitre.org/data/definitions/327.html
+ - https://developer.android.com/reference/javax/crypto/Cipher#getInstance(java.lang.String)
+ - https://developer.android.com/privacy-and-security/security-gms-provider
+ - https://developer.android.com/privacy-and-security/cryptography#bc-algorithms
+ - https://developer.android.com/privacy-and-security/cryptography#jetpack_security_crypto_library
+ - https://developer.android.com/privacy-and-security/cryptography#crypto_provider
+ - https://developer.android.com/privacy-and-security/cryptography#deprecated-functionality
+ - https://www.bsi.bund.de/SharedDocs/Downloads/EN/BSI/Publications/TechGuidelines/TG02102/BSI-TR-02102-1.pdf?__blob=publicationFile
+---
+
+## Overview
+
+The use of outdated or known weak implementations, as well as custom built cryptography poses a significant security risk, as their potential implementation flaws and lack of security review exposes apps to vulnerabilities. Instead, always use certified, expert-designed modules for cryptographic purposes.
+
+## Impact
+
+Using weak or outdated cryptography implementations can lead to a variety of security issues, including but not limited to:
+
+- **Read application data**: Encrypted sensitive data may be compromised by the use of a weak or broken cryptographic algorithm, as it could render it as unencrypted plaintext.
+- **Modify application data**: Integrity-verified application files may be altered due to the use of broken cryptographic algorithms.
+- **Hide sources of data**: Broken cryptographic algorithms may lead to the source of the data of the application to not be proven, if this algorithms are used to ensure the identity of the source of them.
+
+## Modes of Introduction
+
+Weak, outdated or custom cryptographic algorithms can be found in several areas:
+
+- **App Source Code**: In modules where cryptographic algorithms are used.
+- **Libraries**: Third-party or app dependencies where cryptographic algorithms are imported.

```

Comment:
```suggestion
- **Deviation from standard libraries**: Not using well-known libraries for cryptography, such as those provided by the platforms like Conscrypt or CryptoKit, or other well-established libraries like OpenSSL, BouncyCastle, etc.
- **Use of cryptographic constants**: Hardcoded cryptographic constants are typically used to implement cryptographic algorithms. These constants include S-boxes (substitution boxes) for block ciphers, permutation tables, etc.
- **Use of low-level mathematical operations**: Low-level mathematical operations (such as bitwise operations, shifts, custom padding schemes) typically used in cryptographic algorithms.
- **High entropy code**: An indicator of cryptographic implementations or heavily obfuscated code that may hide cryptographic algorithms from reverse engineering.
- **Use of non-cryptographic functions**: Non-cryptographic functions such as Base64 encoding or XOR instead of encryption.
```